tests/websocket.c

```diff
@@ -8,319 +8,537 @@
+    for (char *cur = p; cur < p + sz; cur += 16) {
+        char letters[16] = {0};
+        int limit = (cur + 16) > p + sz ? (sz % 16) : 16;
+        wt->debug("%08lx ", cur - p); /* address */
+        for (int i = 0; i < limit; i++) {
+            wt->debug("%02x ", (unsigned int)(cur[i] & 0xff));
+            letters[i] = isprint(cur[i]) ? cur[i] : '.';
+        }
+        for (int i = limit; i < 16; i++) { /* pad on last line */
+            wt->debug("   ");
+        }
+        wt->debug(" %.*s\n", limit, letters);
+    }
+}
+
+void
+evbuffer_debug_dump(struct worker_thread *wt, struct evbuffer *buffer) {
+    size_t sz = evbuffer_get_length(buffer);
+    char *data = malloc(sz);
+    evbuffer_remove(buffer, data, sz);
+    hex_dump(wt, data, sz);
+    evbuffer_prepend(buffer, data, sz);
+    free(data);
+}
+
+static void
+wait_for_possible_read(struct worker_thread *wt);
+static void
+wait_for_possible_write(struct worker_thread *wt);
+
+static void
+ws_enqueue_frame(struct worker_thread *wt);
+
 void
 process_message(struct worker_thread *wt, size_t sz) {
 
-	// printf("process_message\n");
-	if(wt->msg_received % 10000 == 0) {
-		printf("thread %u: %8d messages left (got %9d bytes so far).\n",
-			(unsigned int)wt->thread,
-			wt->msg_target - wt->msg_received, wt->byte_count);
-	}
-	wt->byte_count += sz;
-
-	/* decrement read count, and stop receiving when we reach zero. */
-	wt->msg_received++;
-	if(wt->msg_received == wt->msg_target) {
-		event_base_loopexit(wt->base, NULL);
-	}
+    // printf("process_message\n");
+    if(wt->msg_received && wt->msg_received % 1000 == 0) {
+        printf("thread %d: %8d messages left (got %9d bytes so far).\n",
+            wt->id,
+            wt->msg_target - wt->msg_received, wt->byte_count);
+    }
+    wt->byte_count += sz;
+
+    /* decrement read count, and stop receiving when we reach zero. */
+    wt->msg_received++;
+    if(wt->msg_received == wt->msg_target) {
+        wt->debug("%s: thread %d has received all %d messages it expected\n",
+            __func__, wt->id, wt->msg_received);
+        event_base_loopexit(wt->base, NULL);
+    }
 }
 
+/**
+ * Called when we can write to the socket.
+ */
 void
-websocket_write(int fd, short event, void *ptr) {
-	int ret;
-	struct worker_thread *wt = ptr;
-
-	if(event != EV_WRITE) {
-		return;
-	}
-
-	char message[] = "\x00[\"SET\",\"key\",\"value\"]\xff\x00[\"GET\",\"key\"]\xff";
-	ret = write(fd, message, sizeof(message)-1);
-	if(ret != sizeof(message)-1) {
-		fprintf(stderr, "write on %d failed: %s\n", fd, strerror(errno));
-		close(fd);
-	}
-
-	wt->msg_sent += 2;
-	if(wt->msg_sent < wt->msg_target) {
-		event_set(&wt->ev_w, fd, EV_WRITE, websocket_write, wt);
-		event_base_set(wt->base, &wt->ev_w);
-		ret = event_add(&wt->ev_w, NULL);
-	}
+websocket_can_write(int fd, short event, void *ptr) {
+    int ret;
+    struct worker_thread *wt = ptr;
+    wt->debug("%s (wt=%p, fd=%d)\n", __func__, wt, fd);
+
+    if(event != EV_WRITE) {
```

Comment:
when would this happen?

Reply:
It doesn't happen since this callback is registered with a single type of event. I have removed this block and the one for reads.